### ISSUE TITLE
Crash with assertion failure when dropping a table with auto_snapshot = false

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#6995]
The following trivial Alternator test creates - using the Alternator API - a single partition with 10,000 items where each has a 10,000 byte string attribute - so totalling a little over 100 MB of data. The data is successfully written, but at the end of the test, the temporary table is deleted, and at that point *then* Scylla crashes with an assertion failure.
The test:
```
def test_big(scylla_only, test_table_sn):
    p = random_string()
    str = 'x' * 10240
    with test_table_sn.batch_writer() as batch:
        for i in range(10000):
            batch.put_item({'p': p, 'c': i, 's': str})
```
The end of the log:
```
INFO  2020-08-07 16:56:27,839 [shard 0] schema_tables - Dropping alternator_alternator_Test_1596808583729.alternator_Test_1596808583729 id=c794a830-d8b5-11ea-9bbe-000000000000 version=b29143f4-29d5-37fd-8c5d-61377619232e
test_scylla: database.cc:1845: database::truncate(const keyspace&, column_family&, database::timestamp_func, bool)::<lambda()>::<lambda()> mutable::<lambda()>::<lambda(db_clock::time_point)>::<lambda()>::<lambda(db::replay_position)>: 
Assertion `low_mark <= rp || rp == db::replay_position()' failed.
Aborting on shard 1.
```
There are no other warnings or errors in the log.

I don't know how Alternator factors into this bug. Alternator doesn't have any special changes to how `database::truncate()` works so it is probably a pre-existing bug in that code. On the other hand, it is possible that something specific in how that test was written - maybe that it made a lot of big writes quickly or never waited for memtable flush or something - exposed this pre-existing bug.

@elcallio "git blame" tells me you write that failing assertion. Can you please help understand what it means and what might cause it to trigger?

**Comment by nyh** [comment]
I discovered that I can only see this assertion failure with "--auto-snapshot 0" (which we use in Alternator's tests). When auto_snapshot is true (which is the default, ), the assertion failure goes away.

**Comment by nyh** [comment]
The difference with `--auto-snapshot 0` which breaks this assertion isn't the actual creation of the snapshot. Rather, it's the fact that with `--auto-snapshot 1`, `database::truncate()` calls `cf.flush()` to flush the last memtable, but when `--auto-snapshot 0`, it calls `cf.clear()` instead because it's pointless to bother flushing the memtable if you intend to immediately delete it anyway. But apparently not flushing the last memtable, or calling `clear()`, causes this assertion to fail.

Maybe our `table::clear()` is implemented incorrectly? It needs to discards memtables immediately, but also to do whatever is appropriate to the commit log to ensure that the assertion on the position, whatever it means, remains true. The current implementation is:
```c++
future<> table::clear() {
    if (_commitlog) {
        _commitlog->discard_completed_segments(_schema->id());
    }
    _memtables->clear();
    _memtables->add_memtable();
    return _cache.invalidate([] { /* There is no underlying mutation source */ });
}
```
@elcallio can you please help me understand if something is wrong in this `clear()` implementation? Or maybe the assertion mentioned above is wrong? 

By the way, in my test, when the assertion fails I see low_mark = {0, 2011303918, 16949256}, rp = {0, 2011303915, 7413768} and db::replay_position() = {0, 0, 0}. I don't know what these numbers mean, and the second numbers change between runs, but low_mark's second number is always 3 higher than rp's second number, and I guess that's why it's considered not less than, so the assertion fails.

**Comment by elcallio** [comment]
Heh. Now that I look at it (and you so helpfully isolated it to `autosnapshot=0`, I can see now that this was in fact broken back in old https://github.com/scylladb/scylla/commit/c2c6c71398470995a739fc35bc5347de8d4148b5, when the condition for flushing was changed from `durable || autosnapshot` to just `autosnapshot`. 

What the former meant was that for all tables that actually had sstables, we did flush + discard. So the assertion holds. _But_ if we suddenly _don't_ flush for tables that actually has sstables on disk, is very much does not. What a silly bunt.

**Comment by nyh** [comment]
But @elcallio, it's actually correct and good that we don't flush the memtables is no autosnapshot! When we intend to delete the table, it's a complete waste of time to write the memtables to disk first... We should just delete them. Doing a `flush()` would be a waste of time, no?
Maybe we should simply not do the `assert()` if `!autosnapshot`? We can just delete the commitlog and everything, and not bother checking anything.

**Comment by elcallio** [comment]
Yes, that is what I mean. The change in question is not bad per sé, but it did not modify the assert to fit the new reality. And I did not manage to see this until now, despite numerous spurious issues popping up over that time. Hello blind spot.

**Comment by nyh** [comment]
> Yes, that is what I mean. The change in question is not bad per sé, but it did not modify the assert to fit the new reality. And I did not manage to see this until now, despite numerous spurious issues popping up over that time. Hello blind spot.

I found 3 potential duplicates for this issue and asked the submitters if perhaps they were using auto_snapshot = false - if they are then it's the same issue. I'm surprised anyone uses Scylla without auto_snapshot = false - what's the point of deleting a table if it stays behind? :-)

**Comment by elcallio** [comment]
Thanks for saving me the work of going through the potential duplicates. Now you volunteered to mark them as such! :-)

**Comment by eliranshlomo1** [comment]
@nyh setting auto_snapshot to true, will prevent the crashes?
If so we might change that in our current production cluster and set a cleanup job for the snapshots

**Comment by elcallio** [comment]
@eliranshlomo1 setting auto snapshot on prevents the bug yes, but we do want both configurations tested. You should probably instead make sure to run scylla versions with the patch applied.

**Comment by nyh** [comment]
> @nyh setting auto_snapshot to true, will prevent the crashes?

Yes, it should (unless we have a second bug in the same area of the code, of course...)

> If so we might change that in our current production cluster and set a cleanup job for the snapshots

Right. At least until you upgrade to a fixed version.

**Comment by nyh** [comment]
This is a fairly annoying bug, that was independently reported at least 4 times (see list of duplicates above), so although it is not a recent regression (it started in March 2019) I think we should backport it. I'll backport to 4.2, 4.1 and 4.0.

**Comment by nyh** [comment]
Backported to next-4.0 (commit 29df4167206ac0a9d058e2bb770f25f2ae79def0), next-4.1 (commit 29df4167206ac0a9d058e2bb770f25f2ae79def0), next-4.2 (commit d70cab044484810cf08d5d0785d491fcce70bbb5)

**Comment by nyh** [comment]
> Backported to next-4.0 (commit [29df416](https://github.com/scylladb/scylla/commit/29df4167206ac0a9d058e2bb770f25f2ae79def0)), next-4.1 (commit [29df416](https://github.com/scylladb/scylla/commit/29df4167206ac0a9d058e2bb770f25f2ae79def0)), next-4.2 (commit [d70cab0](https://github.com/scylladb/scylla/commit/d70cab044484810cf08d5d0785d491fcce70bbb5))

I had a serious mixup here, forgot to backport to 4.1 and then pasted the link to 4.0 and thought I was done with that. So the problem still remains in 4.1 (see #7715) and the fix still needs backporting to 4.1.

**Comment by nyh** [comment]
False alarm... It *was* backported to 4.1 - commit df8d4482c5fef0c75e26e430169342feb39b1704
